messaging/src/main/java/org/axonframework/messaging/annotation/MessageHandlerInterceptorDefinition.java

```diff
@@ -0,0 +1,123 @@
+import org.axonframework.eventhandling.interceptors.MessageHandlerInterceptor;
+import org.axonframework.eventhandling.interceptors.ResultHandler;
+import org.axonframework.messaging.InterceptorChain;
+import org.axonframework.messaging.Message;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * MethodEnhancerDefinition that marks methods (meta-)annotated with {@link MessageHandlerInterceptor}
+ * as interceptors. These methods need to be given special treatment when invoking handlers. Matching
+ * interceptors need to be invoked first, allowing them to proceed the invocation chain.
+ * <p>
+ * This definition also recognizes interceptors only acting on the response. These must be meta-annotated with
+ * {@link ResultHandler}.
+ */
+public class MessageHandlerInterceptorDefinition implements HandlerEnhancerDefinition {
+    @Override
+    public <T> MessageHandlingMember<T> wrapHandler(MessageHandlingMember<T> original) {
+
+        if (original.annotationAttributes(MessageHandlerInterceptor.class).isPresent()) {
+            Optional<Map<String, Object>> attributes = original.annotationAttributes(ResultHandler.class);
+            if (attributes.isPresent()) {
+                return new ResultHandlingInterceptorMember<>(original, (Class<?>) attributes.get().get("resultType"));
+            }
+            return new InterceptedMessageHandlingMember<>(original);
+        }
+        return original;
+    }
+
+    private static class ResultHandlingInterceptorMember<T> extends WrappedMessageHandlingMember<T> implements MessageHandlerInterceptorMember<T> {
+
+        private final Class<?> expectedResultType;
+
+        public ResultHandlingInterceptorMember(MessageHandlingMember<T> original, Class<?> expectedResultType) {
+            super(original);
+            this.expectedResultType = expectedResultType;
+            Method method = original.unwrap(Method.class).orElseThrow(() -> new AxonConfigurationException(
+                    "The @ExceptionHandler must be on method."));
+            boolean declaredInterceptorChain = Arrays.stream(method.getParameters())
+                                                     .anyMatch(p -> p.getType().equals(InterceptorChain.class));
+            if (declaredInterceptorChain) {
+                throw new AxonConfigurationException("@ExceptionHandler annotated methods should not delare an InterceptorChain parameter");
+            }
+        }
+
+        @Override
+        public boolean canHandle(Message<?> message) {
+            return ResultParameterResolverFactory.ignoringResultParameters(() -> super.canHandle(message));
+        }
+
+        @Override
+        public Object handle(Message<?> message, T target) throws Exception {
+            InterceptorChain chain = InterceptorChainParameterResolverFactory.currentInterceptorChain();
+            try {
+                return chain.proceed();
```

Comment:
This is an interceptor that acts on the result of a handler invocation. The proceed call is used to continue in the chain, so that we can invoke the handler "interceptor" with the result of that invocation. In this particular instance, it's a handler that only needs to act on Exceptions, so we catch exceptions and check if the actual exception thrown matches the interceptor's signature. If so, we invoke it, otherwise, we rethrow the exception.

Reply:
Alright, then my gut was right here I feel. Thanks for clarifying further.